send timestamp to database without converting it into UTC

chriskalmar opened the pull request:
fixes #1716

pleerock commented:
What about other drivers?

chriskalmar commented:
If you agree with this approach I'll be more than happy to continue on this for the other drivers. 
I'll just need some time to check how the other engines are handling timezones. MySQL should be an easy one and Oracle should be pretty close to Postgres, don't know about SQL Server though.

pleerock commented:
Can you please review [all issues related to timezone](https://github.com/typeorm/typeorm/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.2.0) I guess this will fix some of them.

chriskalmar commented:
sure, will have a look at them

pleerock commented:
Thank you. Few important notices however:

1. tests are failing for mssql right now
2. please make PR clean so it can be easily reviewed
3. try to keep test as simple as possible, currently it looks like a bit complex, in tests its completely okay to have duplication, the most important is simplicity

chriskalmar commented:
As per request I rewrote all tests and rebased my commits. 
The mssql tests were failing because there is no SQL Server running on travis (ormconfig), so I made each driver test optional based on the current ormconfig. 

I went through all relevant issues and I believe the following will be fixed with this PR:
#1716, #1694, #1240, #976, #838 and #1328 
(possibly even #1189)

pleerock commented:
This change looks like it breaking things (even if it was wrong) and I would like to avoid merging it into the master. What do you think?

chriskalmar commented:
I absolutely understand your concern, but in my experience it's always better to fix problems like that as soon as possible. Of course you have users of typeorm that are already used to (or simply not aware of) the current timestamp/timezone issue and a fix in master would introduce a breaking change for them. 

But on the other hand - due to the growing popularity of typeorm - you have to think of all the new users you get every day and it would be a shame to let them run into problems that could be fixed already. 
PR  #1765 fixes the issue for the next branch, but in my understanding it's not quite ready for prime time. So naturally most users will use the current master releases.

The way I believe it could work out for both, existing and new users, is a compromise:
What if we would introduce a flag or configuration option named e.g. `LEGACY_TIMEZONE_SUPPORT`.
By default typeorm would use the proper timezone handling, but for those that are already having a production system with wrong timezones this flag could bring back the behaviour they are already used to. 

With this approach you enable new users with a correct timestamp/timezone behaviour from day one, and at the same time you provide full backwards compatibility to existing users. And if they decide to upgrade their database to reflect correct timestamps some day it will be their own choice. 

I think everybody would be satisfied. What do you think?

br0wn commented:
I would just note that this now affects `@CreateDateColumn` and  `@UpdateDateColumn` dates as, if I understood correctly, values for those are set with only `new Date()`. Therefore, if the server is running in a timezone, date stored in the database will be stored in the timezone of the server for those columns.

chriskalmar commented:
@br0wn 
`new Date()` creates a timestamp based on UTC. Depending on your timezone you would see a given representation including timezone information. But still, it is UTC.
So the database will store it in UTC as well, which is the correct behaviour.

jwclark commented:
This appears to not be working as intended in 0.2.6 using postgres. The entity and test call below result in a timestamp in server time(rather thanUTC) and without timezone information. 

@Entity()
export class DateTest {

    @PrimaryGeneratedColumn("uuid")
    public id?: string;

    @Column()
    public date?: Date;
}

testRepository.save({date: new Date()});

jwclark commented:
This has been going back and forth since at 466: https://github.com/typeorm/typeorm/issues/466. This was working before.

VinceOPS commented:
Same issue encountered with PostgreSQL
